tensorflow/compiler/tf2tensorrt/convert/utils.h

```diff
@@ -128,77 +129,199 @@ inline bool IsTrtShapeTensorCompatible(const Tensor& tensor) {
-  for (int i = 0; i < dims.nbDims; i++) {
-    dims.d[i] = tensor.dim_size(i);
+  // Construct from an absl::Span.
+  template <typename T>
+  explicit DimsAdapter(absl::Span<T> shape)
+      : num_dims_(static_cast<int32_t>(shape.size())) {
+    absl::c_copy_n(shape, shape.size(), std::back_inserter(storage_));
   }
-  return dims;
-}
 
-template <typename TensorShapeType>
-Status TensorShapeToTrtDims(const TensorShapeType& shape, bool ignore_first_dim,
-                            nvinfer1::Dims* trt_dims) {
-  if (shape.dims() == -1) {
-    trt_dims->nbDims = -1;
+  // Construct from a container.
+  template <typename InputSequence,
+            typename = IsNotTensorShapeType<InputSequence>>
+  explicit DimsAdapter(const InputSequence& shape,
+                       bool ignore_first_dim = false)
+      : DimsAdapter(absl::MakeSpan(shape).subspan(ignore_first_dim ? 1 : 0,
+                                                  shape.size())) {}
+
+  // Construct from a status.
+  explicit DimsAdapter(Status status) : status_(std::move(status)) {}
+
+  //----- Conversion Utilities ------
+
+  // Convert to an nvinfer1::Dims and assign to the passed pointer.
+  Status TrtDims(nvinfer1::Dims* result) const {
+    TF_RETURN_IF_ERROR(status_);
+    result->nbDims = num_dims_;
+    absl::c_copy(storage_, static_cast<int32_t*>(result->d));
     return Status::OK();
   }
-  return ContainerToTrtDims(shape.dim_sizes(), trt_dims, ignore_first_dim);
-}
 
-Status GetNetworkInputShapes(const nvinfer1::INetworkDefinition* network,
-                             std::vector<PartialTensorShape>* input_shapes);
+  // Convert to an nvinfer1::Dims and return by value.
+  StatusOr<nvinfer1::Dims> AsTrtDims() const {
+    nvinfer1::Dims result;
+    TF_RETURN_IF_ERROR(TrtDims(&result));
+    return result;
+  }
 
-// Creates PartialTensorShape from nvinfer1::Dims. The "batch_dimension", when
-// provided, is prepended to the shape.
-inline PartialTensorShape TrtDimsToPartialTensorShape(
-    const nvinfer1::Dims& dims,
-    absl::optional<int64> batch_dimension = absl::nullopt) {
-  absl::InlinedVector<int64, 4> dims_vec;
-  if (batch_dimension) {
-    dims_vec.insert(dims_vec.begin(), std::max(int64(-1), *batch_dimension));
-  }
-  std::copy(dims.d, dims.d + dims.nbDims, std::back_inserter(dims_vec));
-  return PartialTensorShape(dims_vec);
-}
+  // Convert to a TensorShape.
+  Status TensorShape(TensorShape* shape,
+                     absl::optional<int> batch_size = absl::nullopt) const {
+    TF_RETURN_IF_ERROR(status_);
+    TF_RETURN_IF_ERROR(TensorShapeUtils::MakeShape(
+        static_cast<const int64_t*>(storage_.data()), storage_.size(), shape));
+    if (batch_size) shape->InsertDim(0, *batch_size);
+    return Status::OK();
+  }
 
-Status TrtDimsToTensorShape(const std::vector<int>& trt_dims,
-                            TensorShape* shape,
-                            absl::optional<int> batch_size = absl::nullopt);
+  Status PartialTensorShape(
```

Comment:
Let's add this comment:

Similar to TensorShape above, but for PartialTensorShape.

Reply:
Also updated vector to vector<T> where T is integer